src/test/groovy/graphql/normalized/ExecutableNormalizedOperationToAstCompilerTest.groovy

```diff
@@ -152,17 +149,650 @@ class ExecutableNormalizedOperationToAstCompilerTest extends Specification {
+        }
+        """)
+
+        String query = """
+        {
+            animal {
+                parent {
+                    name
+                }
+            }
+        }
+        """
+
+        def tree = createNormalizedTree(schema, query)
+        // printTreeWithLevelInfo(tree, schema).forEach { println it }
+
+        when:
+        def result = compileToDocument(schema, QUERY, null, tree.topLevelFields, noVariables)
+        def printed = AstPrinter.printAst(new AstSorter().sort(result.document))
+
+        then:
+        printed == """query {
+  animal {
+    parent {
+      name
+    }
+  }
+}
+"""
+    }
+
+    def "test interface fields with different output types on the implementations 2"() {
+        def schema = TestUtil.schema("""
+        interface Animal {
+            parent: Animal
+            name: String
+        }
+        type Cat implements Animal {
+            name: String
+            parent: Cat
+        }
+        type Dog implements Animal {
+            name: String
+            parent: Dog
+            isGoodBoy: Boolean
+        }
+        type Query {
+            animal: Animal
+        }
+        """)
+
+        String query = """
+        {
+            animal {
+                parent {
+                    name
+                    ... on Dog {
+                        isGoodBoy
+                    }
+                }
+            }
+        }
+        """
+
+        def tree = createNormalizedTree(schema, query)
+        // printTreeWithLevelInfo(tree, schema).forEach { println it }
+
+        when:
+        def result = compileToDocument(schema, QUERY, null, tree.topLevelFields, noVariables)
+        def printed = AstPrinter.printAst(new AstSorter().sort(result.document))
+
+        then:
+        printed == """query {
+  animal {
+    parent {
+      name
+      ... on Dog {
+        isGoodBoy
+      }
+    }
+  }
+}
+"""
+    }
+
+    def "test interface fields with different output types on the implementations 3"() {
+        // Tests we don't consider File as a possible option for parent on animals
+        def schema = TestUtil.schema("""
+        interface Node {
+            parent: Node
+        }
+        type File implements Node {
+            name: ID
+            parent: File
+        }
+        interface Animal implements Node {
+            parent: Animal
+            name: String
+        }
+        type Cat implements Animal & Node {
+            name: String
+            parent: Cat
+        }
+        type Dog implements Animal & Node {
+            name: String
+            parent: Dog
+            isGoodBoy: Boolean
+        }
+        type Query {
+            animal: Animal
+            file: File
+        }
+        """)
+
+        String query = """
+        {
+            animal {
+                parent {
+                    name
+                    ... on Dog {
+                        isGoodBoy
+                    }
+                    ... on Node {
```

Comment:
this is a great test. We should probably have another one where this is a standalone fragment and not inlined and used in another context where `...on Node` actually can include `File`